### PR TITLE
Properly enforce the pathLenConstraint of the BasicConstraints extension

### PR DESCRIPTION
**Pull request by julek-wolfssl**
- move the testsuite file helpers into a new `tests/utils.c` file so that they can be used across all tests
- dump the raw TLS stream when WOLFSSL_DUMP_MEMIO_STREAM is defined so that it can be examined in Wireshark

Addresses ZD15258

### CODE REVIEW

**Review by dgarske — DISMISSED** [review]
Also asking Sean to review

File: wolfcrypt/src/asn.c [hunk 1]
```diff
@@ -22614,93 +22629,32 @@ int ParseCertRelative(DecodedCert* cert, int type, int verify, void* cm)
...
-                if (decrementMaxPathLen && cert->ca->maxPathLen > 0) {
-                    WOLFSSL_MSG("\tmaxPathLen status: reduce by 1");
-                    cert->maxPathLen = (byte)(cert->ca->maxPathLen - 1);
-                    if (verify != NO_VERIFY && type != CA_TYPE &&
-                                                    type != TRUSTED_PEER_TYPE) {
-                        WOLFSSL_MSG("\tmaxPathLen status: OK");
-                    }
-                } else if (decrementMaxPathLen && cert->ca->maxPathLen == 0) {
+        if (!cert->selfSigned) {
+            if (/* Need to perform a pathlen check on anything that will be used
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c:22637]: Don't mess the if statement up with comments.
Move the comment to above if statement.

File: src/internal.c [hunk 2]
```diff
@@ -12654,7 +12654,8 @@ void DoCertFatalAlert(WOLFSSL* ssl, int ret)
     alertWhy = bad_certificate;
     if (ret == ASN_AFTER_DATE_E || ret == ASN_BEFORE_DATE_E) {
         alertWhy = certificate_expired;
-    } else if (ret == ASN_NO_SIGNER_E) {
+    } else if (ret == ASN_NO_SIGNER_E || ret == ASN_PATHLEN_INV_E ||
```
Comment (SparkiDev) [inline src/internal.c:12657]: Fix formatting.

File: src/ssl.c [hunk 3]
```diff
@@ -9430,6 +9433,10 @@ int wolfSSL_CertManagerLoadCA(WOLFSSL_CERT_MANAGER* cm, const char* file,
         return ret;
     }
 
+    /* Some configurations like OPENSSL_COMPATIBLE_DEFAULTS may turn off
+     * verification by default. Let's restore our desired defaults. */
+    wolfSSL_CTX_set_verify(tmp, WOLFSSL_VERIFY_DEFAULT, NULL);
```
Comment (SparkiDev) [inline src/ssl.c:9438]: We are using a temporary WOLFSSL_CTX to load certificate.
Are the defaults not set by default?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 3 above)
Comment (julek-wolfssl) [reply src/ssl.c:9438]: When `OPENSSL_COMPATIBLE_DEFAULTS` is defined, then verification gets turned off in `wolfSSL_CTX_new_ex`. I left this in for all configurations to make sure that nothing turns off verification as that is what we want in this function.

File: wolfcrypt/src/asn.c (hunk 1 above)
Comment (julek-wolfssl) [reply wolfcrypt/src/asn.c:22637]: Moving

File: src/internal.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/internal.c:12657]: Fixing.

File: wolfcrypt/src/asn.c (hunk 1 above)
Comment (julek-wolfssl) [reply wolfcrypt/src/asn.c:22637]: Moved comments.

File: src/internal.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/internal.c:12657]: Fixed.

**Comment by julek-wolfssl** [comment]
Retest this please.

**Review by SparkiDev — APPROVED** [review]

**Comment by kojo1** [comment]
I have confirmed additional test cases by the customer. Thanks!

**Review by kojo1 — APPROVED** [review]

**Comment by julek-wolfssl** [comment]
retest this please.

**Review by dgarske — APPROVED** [review]

**Pull request merged**